application/src/Controller/Admin/VocabularyController.php

```diff
@@ -115,26 +125,112 @@ public function editAction()
         $data = $vocabulary->jsonSerialize();
         $form->setData($data);
 
+        $view = new ViewModel;
+        $view->setVariable('vocabulary', $vocabulary);
+
         if ($this->getRequest()->isPost()) {
             $data = $this->params()->fromPost();
             $form->setData($data);
             if ($form->isValid()) {
                 $response = $this->api($form)->update('vocabularies', $this->params('id'), $data, [], true);
                 if ($response->isSuccess()) {
-                    $this->messenger()->addSuccess('Vocabulary successfully updated'); // @translate
-                    return $this->redirect()->toRoute(null, ['action' => 'browse'], true);
+                    $fileData = $this->params()->fromFiles('file');
+                    if (0 === $fileData['error']) {
+                        $this->messenger()->addSuccess('Vocabulary label and comment successfully updated. Please review these changes before you accept them.'); // @translate
+                        $diff = $this->rdfImporter->getDiff(
+                            'file',
+                            $vocabulary->namespaceUri(),
+                            ['file' => $fileData['tmp_name']]
+                        );
+                        $form = $this->getForm(VocabularyUpdateForm::class);
+                        $form->setAttribute('action', $this->url()->fromRoute(null, ['action' => 'update'], true));
+                        $form->get('diff')->setValue(json_encode($diff));
+
+                        $view->setVariable('diff', $diff);
+                        $view->setTemplate('omeka/admin/vocabulary/update');
+                    } else {
+                        $this->messenger()->addSuccess('Vocabulary successfully updated'); // @translate
+                        return $this->redirect()->toRoute(null, ['action' => 'browse'], true);
+                    }
                 }
             } else {
                 $this->messenger()->addFormErrors($form);
             }
         }
 
-        $view = new ViewModel;
-        $view->setVariable('vocabulary', $vocabulary);
         $view->setVariable('form', $form);
         return $view;
     }
 
+    public function updateAction()
```

Comment:
We talked about this before you did it, but now seeing it, I don't know how happy I am with the decision to do the updating within the controller.

Minimally I'd prefer the "actually update given this diff" code to live in the Importer and just be called from here.